audit/audit.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 Mia srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package audit
+
+import (
+	"reflect"
+	"slices"
+
+	"github.com/google/uuid"
+)
+
+const (
+	AuditAdditionalDataGrantedPermissionKey = "authorization.permission"
+	AuditAdditionalDataGrantedBindingKey    = "authorization.binding"
+	AuditAdditionalDataGrantedRoleKey       = "authorization.role"
+)
+
+var reservedLabelKeys = []string{
+	AuditAdditionalDataGrantedPermissionKey,
+	AuditAdditionalDataGrantedBindingKey,
+	AuditAdditionalDataGrantedRoleKey,
+}
+
+type auditReservedFields struct {
+	ID string `audit:"id"`
+}
+
+type Audit struct {
+	auditReservedFields
+	AggregationID string                 `audit:"aggregationId"`
+	Authorization AuthzInfo              `audit:"authorization"`
+	Subject       SubjectInfo            `audit:"subject"`
+	RequestBody   interface{}            `audit:"requestBody"`
+	Labels        map[string]interface{} `audit:"labels"`
+}
+
+type authzinfoReservedFields struct {
+	GrantingPermission string `audit:"permission"`
+	GrantingBindingID  string `audit:"binding"`
+	GrantingRoleID     string `audit:"roleId"`
+}
+
+type AuthzInfo struct {
+	Allowed    bool   `audit:"allowed"`
+	PolicyName string `audit:"policyName"`
+	authzinfoReservedFields
+}
+
+type SubjectInfo struct {
+	ID     string   `audit:"id"`
+	Groups []string `audit:"groups"`
+}
+
+func (a *Audit) generateID() {
+	a.auditReservedFields.ID = uuid.NewString()
+}
+
+func (a *Audit) applyDataFromPolicy(data map[string]interface{}) {
```

Comment:
I'd use map[string]any instead (also for the toMap function). Furthermore, why not set directly in a map in the `applyDataFromPolicy`? In this way there are 2 conversions, but are the data used?

Reply:
I've changed the code to use the fancier `map[string]any`. However I don't see the two conversion you mention 🤔 

I get a map[string]any within the built-in function and cache it, then I extract it from the cache as-is and apply it on the `Audit struct`.

If by this 

> why not set directly in a map

you mean: why having the Audit struct to begin with, it is meant to provide a structured format to the sdk that uses  the Trace function. If I'd accept there a map[string]any then there could be typos in the map creation that lead to bad audit being generated.


Consider that the `toMap` function is only used to convert the Audit struct to a map to be properly supplied to the logging.Fields type, therefore for stdout logging only, when moving to db/sink-based auditing we won't be using the tomap function anymore :)

Reply:
Ok, I misunderstood what the `data` is here. 

No, it's ok to set the audit as a struct. But we want the Audit struct fixed for the audit itself, or decided by the sdk? 
Anyhow, ok with it

Reply:
at the moment I kept it in the audit package, I guess the SDK could provide some primitives to extend the data, not the whole audit 🤔

Reply:
I've made a few changes to this as the previous impl had some issues with the "reserved" fields substructures